services/report/__init__.py

```diff
@@ -719,6 +722,30 @@ async def create_new_report_for_commit(self, commit: Commit) -> Report:
                 paths_to_carryforward,
                 session_extras=dict(carriedforward_from=parent_commit.commitid),
             )
+            # If the parent report has labels we also need to carryforward the label index
+            # Considerations:
+            #   1. It's necessary for labels flags to be carryforward, so it's ok to carryforward the entire index
+            #   2. As tests are renamed the index might start to be filled with stale labels. This is not good.
+            #      but I'm unsure if we should try to clean it up at this point. Cleaning it up requires going through
+            #      all lines of the report. It might be better suited for an offline job.
```

Comment:
I agree - could this happen in the upload finisher perhaps **after** triggering the notification task?

Reply:
I'm unsure... certainly after the notification task. But I'm not sure about the upload finisher because it is scheduled for every `upload` task. It does have a lock for it, but maybe I'll have to implement a multi-reader/single write lock for the labels file?

My biggest concern with putting it in the upload finisher is that there might still be uploads being actively processed at the same time. I think the cleanup should happen a few hours after the last upload came in, to be extra sure.